Add formatter for the parse AST

LaurenzV opened the pull request.

LaurenzV committed:
First version of reformatter

LaurenzV committed:
Don't translate macros when reformatting

LaurenzV committed:
Refine expressions further

LaurenzV committed:
Update comment

LaurenzV committed:
Fix spacing

LaurenzV committed:
Change PForAll

LaurenzV committed:
more

LaurenzV committed:
Tweak formatting of let and if statements

LaurenzV committed:
First version of better trivia fetching

LaurenzV committed:
More broken now but still more progress

LaurenzV committed:
Make reformatter context implicit

LaurenzV committed:
Remove unused method

LaurenzV committed:
roll back to looking back for trivia

LaurenzV committed:
Fix issue with missing indent after comment

LaurenzV committed:
Everything works as initially

LaurenzV committed:
Simplify code a bit

LaurenzV committed:
next attempt

LaurenzV committed:
Good progress towards better formatting!

LaurenzV committed:
Fix spacing in pre and post conditions

LaurenzV committed:
Improve showBody method

LaurenzV committed:
more

LaurenzV committed:
Create ReformatterPrettyPrinterBase

LaurenzV committed:
More progress

LaurenzV committed:
more progress towards new reformatter

LaurenzV committed:
First seeminly working prototype

LaurenzV committed:
First working version, more or less back to where we started

LaurenzV committed:
Add short circuit

LaurenzV committed:
Add hack for decreases keyword

LaurenzV committed:
Only reformat on leaf nodes

LaurenzV committed:
Fix trailing whitespaces

LaurenzV committed:
Fix whitespace bug

LaurenzV committed:
Add a couple of tests for reformatting

LaurenzV committed:
Remove POther

LaurenzV committed:
Add a test case for not working snippets

LaurenzV committed:
Add some license headers

LaurenzV committed:
Revert some unnecessary changes

LaurenzV committed:
Remove trailing whitespaces

LaurenzV committed:
Remove parseResult variable

LaurenzV committed:
Whitespaces and other small changes

LaurenzV committed:
Rearrange some imports

LaurenzV committed:
Rename trivia to PTrivia

LaurenzV committed:
Remove LeftLineIndent

LaurenzV committed:
Readd newlines

LaurenzV committed:
Add some comments and improve code formatting

LaurenzV committed:
Add another case for showAny

LaurenzV committed:
Fix whitespaces and imports

LaurenzV committed:
Add brackets to method calls

LaurenzV committed:
Fix some whitespace issues and similar

LaurenzV commented:
@rayman2000

LaurenzV committed:
Merge branch 'master' into format-rnode

LaurenzV committed:
Merge remote-tracking branch 'mine/format-rnode' into format-rnode

LaurenzV committed:
Move tests from viperserver to silver

LaurenzV committed:
Merge branch 'master' into format-rnode

JonasAlaif commented on src/main/scala/viper/silver/parser/FastParser.scala:
```diff
@@ -167,7 +187,7 @@ object FastParserCompanion {
 
 class FastParser {
-  def parse(s: String, f: Path, plugins: Option[SilverPluginManager] = None, loader: FileLoader = DiskLoader) = {
+  def parse(s: String, f: Path, expandMacros: Boolean, plugins: Option[SilverPluginManager] = None, loader: FileLoader = DiskLoader) = {
```
Let's make this the last argument and add the default of `true` here to avoid changing the api.

JonasAlaif commented on src/main/scala/viper/silver/parser/FastParser.scala:
```diff
@@ -104,6 +104,26 @@ object FastParserCompanion {
+
+  def comment[$: P]: P[PComment] = lineComment | blockComment
+
+  def trivia[$: P]: P[Seq[PTrivia]] = {
```
Pick a better name (and rename everywhere). Did you mean `trivial`?

LaurenzV replied:
I've seen `trivia` used in other projects before where it refers to whitespaces + comments, that's why I named it this way, but I can try to come up with something different.

LaurenzV replied:
Would just `commentsAndWhitespaces` work for you?

JonasAlaif replied:
> I've seen trivia used in other projects before where it refers to whitespaces + comments

Ah cool, I hadn't seen that before and thought it was just a typo. My bad, it's fine to keep it as it is then.

JonasAlaif commented on src/main/scala/viper/silver/parser/FastParser.scala:
```diff
@@ -104,6 +104,26 @@ object FastParserCompanion {
+  def newline[$: P]: P[PNewLine] = P((StringIn("\n\r") | "\n" | "\r") map (_ => PNewLine()))
+
+  def lineComment[$: P]: P[PComment] = {
+    P(("//" ~~ CharsWhile(_ != '\n').?.!).map { content =>
```
Use these two in the `implicit val whitespace` defn so that they stay in sync.

LaurenzV replied:
Done, I also did it for spaces and newlines. The only thing I had to change is that the parser now also treats `\r\n` as one single newline, otherwise you might get double linebreaks on Windows after reformatting.

JonasAlaif commented on src/main/scala/viper/silver/parser/ParseAst.scala:
```diff
@@ -8,8 +8,17 @@ package viper.silver.parser
+import viper.silver.ast.utility.rewriter.{HasExtraValList, HasExtraVars, Rewritable, StrategyBuilder}
+import viper.silver.ast.{Exp, Member, NoPosition, SourcePosition, Stmt, Type, FilePosition, HasLineColumn, Position}
+import viper.silver.parser.ReformatPrettyPrinter.{show, showAnnotations, showAny, showBody, showInvs, showOption, showPresPosts, showReturns, showSeq}
+import viper.silver.parser.PSymOp.{EqEq, Iff, Implies}
```
Keep the import formatting from above (i.e. in one line if possible)

LaurenzV replied:
I changed it now so that all methods are in a single object `RNode` so that you can do `import viper.silver.parser.RNode._`, I hope that's what you meant.

JonasAlaif commented on src/main/scala/viper/silver/parser/ParseAst.scala:
```diff
@@ -231,7 +240,9 @@ trait PIdentifier extends PLeaf {
 
-case class PIdnDef(name: String)(val pos: (Position, Position)) extends PNode with PIdentifier
+case class PIdnDef(name: String)(val pos: (Position, Position)) extends PNode with PIdentifier {
+  override def reformat(implicit ctx: ReformatterContext): List[RNode] = rt(name)
```
Move to `PIdentifier` and remove from other classes implementing the trait. Is it not possible to reuse the `PLeaf` and its `display` function to avoid `reformat` implementation for all classes implementing `PLeaf`?

JonasAlaif commented on src/main/scala/viper/silver/parser/ParseAst.scala:
```diff
@@ -318,7 +335,9 @@ case class PVersionedIdnUseExp(name: String, version: String, separator: String
 /** The declaration of an argument to a domain function. Not a `PDeclaration` as it will never clash. */
-case class PDomainFunctionArg(name: Option[PIdnDef], c: Option[PSym.Colon], typ: PType)(val pos: (Position, Position)) extends PAnyFormalArgDecl
+case class PDomainFunctionArg(name: Option[PIdnDef], c: Option[PSym.Colon], typ: PType)(val pos: (Position, Position)) extends PAnyFormalArgDecl {
+  override def reformat(implicit ctx: ReformatterContext): List[RNode] = showOption(name) <> showOption(c) <+> show(typ)
```
I really have a feeling we could avoid a lot of these functions by simply using the `PPrettySubnodes` trait (and e.g. adding a `reformat` function there or even just merge `pretty` and `reformat`).

LaurenzV replied:
So I've gotten rid of quite a few implementations in https://github.com/viperproject/silver/pull/820/commits/ecbb283c561c88a798873a4b5d4278b59c37b243, let me know if it is better now. There are still quite a few implementations, but they are kind of unavoidable because a lot of nodes need custom handling.

I can try to move the `Reformattable` trait into PPrettySubNodes. I'm not a 100% sure about merging `pretty` and `reformat`, since I'm not sure if someone maybe relies on the API and I want to avoid breaking/interfering with existing code as much as possible. Not to mention that it would result in an even bigger git diff, so I think if at all it would make sense to do it in a separate PR? As you want, though!

JonasAlaif replied:
Sorry for the slow response, I was on holiday. `pretty` is only used in the IDE e.g. to display the function signature of a function call when hovered: I imagine that it should match quite closely to what reformat does. For example, `pretty` is already able to handle printing this `PDomainFunctionArg` - the `PSym.Colon` itself adds the space. I imagine `reformat` could be made to work the same way, with little custom handling here.

LaurenzV replied:
So just to be clear, you still want me to replace `pretty` completely in this PR? Or would it be fine to open an issue for that and for now, I just focus on trying to merge this into `PPrettySubNodes` and seeing whether I can remove some more impls?

JonasAlaif replied:
I would like this PR to do things the way that `pretty` does. That means that the vast majority of classes in `ParseAst.scala` don't have a `pretty` or `reformat` implementation (looking over it now, I even see some overrides of `pretty` that are unnecessary), except for the unparseable types and 2 or 3 special pnodes maybe. If you're not sure how to achieve this, I'm happy to help out a bit.

If `pretty` is deleted once that is done is not super important. I imagine it would simply be implemented as `.reformat.toString` or whatever the calls are.

LaurenzV replied:
Got it, I’ve already managed to remove a couple more overrides, I will keep trying a few more things and see how far I can get.

JonasAlaif commented on src/main/scala/viper/silver/plugin/ParserPluginTemplate.scala:
```diff
@@ -6,13 +6,13 @@
 import viper.silver.verifier.VerificationResult
 
 import scala.collection.Set
-
```
Avoid arbitrary line changes like this (also other cases)

JonasAlaif commented on src/test/scala/SemanticAnalysisTests.scala:
```diff
@@ -31,7 +31,7 @@ class SemanticAnalysisTests extends AnyFunSuite {
     val body = PSeqn(PDelimited.impliedBlock(Seq(PAssert(PReserved.implied(PKw.Assert), binExp1)(p), PSeqn(PDelimited.impliedBlock(Seq(PAssert(PReserved.implied(PKw.Assert), binExp2)(p))))(p))))(p)
     val method = PMethod(Seq(), PReserved.implied(PKw.Method), PIdnDef("m")(p), PGrouped.impliedParen(PDelimited.empty), None, PDelimited.empty, PDelimited.empty, Some(body))(p)
-    val program = PProgram(Nil, Seq(method))(p, Seq())
+    val program = PProgram(Nil, Seq(method))(p, Seq(), Seq(),"")
```
`Seq(),""` -> `Seq(), ""` (same issue below)

JonasAlaif reviewed (changes requested):
I think that we really don't need all of these `reformat` override defns. Please find some way to avoid most of them (see my suggestion for how).

LaurenzV committed:
Avoid arbitrary whitespace changes

LaurenzV committed:
Fix whitespaces in PProgram constructors

LaurenzV committed:
Make expandMacros the last argument and assign true by default

LaurenzV committed:
Avoid another arbitrary whitespace change

LaurenzV committed:
Fix compilation

LaurenzV committed:
Deduplicate comment and whitespace parsing

LaurenzV committed:
Reduce the number of imports

LaurenzV committed:
Fix compiler warnings

LaurenzV committed:
Improve whitespace handling again

LaurenzV committed:
Deduplicate other whitespace, too

LaurenzV committed:
Fix wrong expandMacros

LaurenzV committed:
Undo newline

LaurenzV committed:
Merge branch 'master' into format-rnode

LaurenzV committed:
Reduce the number of `reformat` and `reformatExp` implementations

LaurenzV committed:
Remove one more reformat override

LaurenzV committed:
More improvements

JonasAlaif commented on src/main/scala/viper/silver/parser/ParseAstKeyword.scala:
```diff
@@ -19,6 +22,8 @@ trait LeftSpace extends PReservedString { override def leftPad = " " }
 case class PReserved[+T <: PReservedString](rs: T)(val pos: (Position, Position)) extends PNode with PLeaf {
   override def display = rs.display
+  // Need to override implementation because pretty-printing will add space padding
+  override def reformat(implicit ctx: ReformatterContext): List[RNode] = rt(rs.token)
```
I would change `leftPad` and `rightPad` to be Boolean (I think currently they only are `""` or `" "`) and then add a space to the left/right based on that. I think you can do the majority of spaces this way. You could also add a `addNewline` field which is set for e.g. all stmt keywords and `function`/`method`/`requires` etc. which adds a newline before the keyword. Lastly, I would define a some things as "grouping" in the sense that they add two spaces after a newline for all contained `RNode`s, which would probably handle most cases of indentation (I think you already do this below, but I'm not 100% sure).

LaurenzV committed:
Fix issue with empty assignments

LaurenzV committed:
Fix inhale exhale exp

LaurenzV committed:
Add manual override for package wand

LaurenzV commented:
So, overall, I think I've managed to get rid of a bunch of overrides, thanks for the hints, they helped a lot! :D However, there still are quite a few remaining but I think it will be rather hard to remove those... :/ I'll try to add explanations for some of them below.

Some other notes:
Unfortunately, I couldn't reuse `leftPad`/`rightPad` in all instances, because there are a few examples where we need to deviate from the pretty printer (for example, the pretty printer by default wraps any binary operator with spaces, but for the reformatter we handle operators like `+` and `==>` differently, so always padding with spaces messes things up). So I took the route of using the padding of the pretty printer by default, but override it for a few select cases.

Regarding some of the overridden `reformat`s:
`PGrouped`, `PDelimited`: As you can see in the code there's quite a few different cases that need to be distinguished, so I don't think this can easily be removed.

`PFunctionType`: We can't reuse the `pretty` impl here because it would call `pretty` on all children as well.

`PExp`/`PBinExp`: For those we also have different code paths depending on what exactly we are processing, so it's again hard to find a generalization for these.

`PCondExp`,`PUnfolding`,`PApplying`,`PAsserting`, etc.: For those (i.e. the ones with PExp), the difficulty is that the nesting we apply does not really have a clear general pattern, so I'm not exactly sure how you would generically apply the grouping. For example, for `unfolding` you have a single expression which is nested, grouped and then prepended with a line (which will evaluate to either a space or a new line depending on the situation), while for `PCondExp`, we have two expressions, out of which the else part shouldn't get an additional layer of nesting. Also, always prepending a line to an expression is also not desirable, for example, if you have a field access somewhere in your code, this shouldn't automatically have a prepended whitespace.

`PIf`, `PWhile`, etc.: Those also have somewhat conditional formatting. For example, if a while loop has invariants, then the opening bracket should always be on a newline, while if there aren't any invariants it looks nicer if it's on the same line.

`PDefine`: I don't think I can remove the manual space here since adding spaces to either `PDelimited` or even `PNode` will break a lot of other things.

`PDomainFunction`, `PExp`: Same here, adding a space by default to either `PIdnDef` or `PDelimited` or `PBracedExp` would break lots of other things.

Similar problems apply to the remaining ones.

Regarding your comments:

> I would change leftPad and rightPad to be Boolean (I think currently they only are "" or " ") and then add a space to the left/right based on that.

I changed that!

> You could also add a addNewline field which is set for e.g. all stmt keywords and function/method/requires etc. which adds a newline before the keyword.

I tried this, unfortunately this doesn't seem to work very well in my case... The problem is that I had to implement a somewhat sophisticated logic to preserve newlines in the existing program while also respecting the intended formatting, and it doesn't work well if I just always prepend a newline. Just to give you an idea, if you have for example:

```
field x: Int
field y: Int
```

After reformatting, it should preferably stay the same, while if you have something like:

```
field y: Int



predicate StructA(this: Ref) {
  acc(this.x) && acc(this.y)
}
```

The reformatted result will now look like:
```
field y: Int

predicate StructA(this: Ref) {
  acc(this.x) && acc(this.y)
}
```

instead of:
```
field y: Int
predicate StructA(this: Ref) {
  acc(this.x) && acc(this.y)
}
```

Things get even more complicated if there are comments in between the members where I need to look at the whitespaces before and after the outer comments, so this is why the logic for reformatting is pretty delicate, and just prepending a newline before each members unfortunately trips it up.

> Lastly, I would define a some things as "grouping" in the sense that they add two spaces after a newline for all contained RNodes, which would probably handle most cases of indentation (I think you already do this below, but I'm not 100% sure).

Yes I have the `rne()` function which will indent everything that appears in-between, but as mentioned above there are kind of a lot of edge cases with different constructs having different preferred nesting rules, so generalizing those doesn't seem easy to me... The reason the pretty printer has much less of those is that it simplifies a lot, and therefore the output isn't great in many cases and often even produces syntactically broken results (I tried it). But for reformatting, it's obviously important that the output looks nice and the program stays syntactically valid.

Let me know what you think, I hope it's at least an improvement to before.

rayman2000 commented:
@LaurenzV Jonas and I discussed this and we think we should not remove the pretty() calls (which would break ViperServer) but wherever possible, have pretty() just call reformat() internally so that we do not have to parallel implementations of almost the same thing. Could you take a look at that?

LaurenzV commented:
That's already what happens, see here: https://github.com/LaurenzV/silver/blob/063279a982dfd0d1c2af49c0e009610f290e7500/src/main/scala/viper/silver/parser/ParseAst.scala#L41-L47. 

All of the remaining overrides are (as mentioned above) more complex and thus I'm not really sure how I can remove them without sacrificing the output quality of the reformatter.

LaurenzV commented:
Ah, sorry. I thought you meant `reformat` should call `pretty`, not the other way around. I can look into that!

rayman2000 commented:
Ah sorry, I misunderstood that. It looks like reformat calls pretty and not the other way around? But that is also fine. So then this is ready from your point of view?

LaurenzV commented:
Yes, right now, `reformat` will call `pretty` as a fallback. I can of course try letting `pretty` call `reformat` instead, though it might be a bit tricky because the reformatter also expects a reformatter context with information about comments/whitespaces in the program, which doesn't really exist if we pretty print a node individually. So without trying it, I can't tell how feasible it is.

Otherwise, this would be ready from my side. :)

rayman2000 commented:
Yeah I think you can leave it as is, the main goal is to not have duplicate stuff, the direction does not matter too much to me.
@JonasAlaif can you take another look

JonasAlaif committed:
Simplify and remove some reformat overrides

JonasAlaif commented:
@LaurenzV see my last commit. I've changed things to be even more in the direction that I suggested. The expected formatting changed a tiny bit (check that it looks reasonable, or revert it as you wish). There is still one failing test that I couldn't figure out - maybe you could have a look.

If there are parts you don't like about this version, I'm happy to discuss alternative solutions :)

LaurenzV committed:
Don't group expressions

LaurenzV committed:
Try to fix CI failure

LaurenzV committed:
Revert "Try to fix CI failure"

This reverts commit ae824da99471399ebd9a5f1b6a731a12bae97e8f.

LaurenzV commented:
Thanks a lot! Seems like you had to change a lot, sorry about that, I didn't have any experience with Scala or Viper before, so I imagine a lot of the code was unidiomatic. 😅 

I have fixed the issue with the failing silver test, though it seems like your changes to the parsing and macro expansion logic are causing CI failures, I thought [this](https://github.com/viperproject/silver/pull/820/commits/ae824da99471399ebd9a5f1b6a731a12bae97e8f) would fix it, but looks like it doesn't fully solve the issue, so I've reverted it for now. Any idea what the best solution to the problem is? I can try to look into it again tomorrow or so, but given that I'm not at all familiar with that part of the code base, I'm not sure I would be able to figure out the right solution fast.

Other then that, I don't have any objections to your changes, especially since the test cases still pass, which do have a pretty wide coverage.

LaurenzV commented:
Since some of them are match errors, it might mean we need to make changes in the Silicon test suite? Not sure if that's acceptable to change, though.

JonasAlaif commented:
I think that most of the errors are stemming from the change to `PIdnUseExp`, it used to be both an `PExp` and `PIdnUseName`. I changed it to be a `PExp` only which wraps a `PIdnUseName`, since then there aren't two conflicting implementations of `reformat` (i.e. the `PLeaf` one from `PIdnUseName` and the `PExp` one). I think the change makes sense, but there are various places which match a `_: PIdnUseName` expecting the `PIdnUseExp` to match, but it no longer does. I think that this is likely the cause for both the match errors and the macro expansion errors. I tried to change the occurrences of this match in the macro expansion code, but it seems that I must have missed something.

I couldn't find the parse error. Could you send a link to the CI of that one?

Regarding changing the SIlicon test suite, it's a bit annoying but definitely in scope for this PR.

LaurenzV commented:
> I couldn't find the parse error. Could you send a link to the CI of that one?

Ah no, I didn't mean that there is a specific error related to parsing, just that as part of those changes the CI failures appeared. It's possible that they are all just related to the `PIdnUseExp` thing.

Okay, in this case I will give it another shot a bit later and try to fix the failures. Thanks again!

LaurenzV committed:
Revert "Revert "Try to fix CI failure""

This reverts commit 053cdde404ffe511616699320cf4e06345802e05.

LaurenzV committed:
Merge branch 'master' into format-rnode

LaurenzV committed:
Maybe fix Carbon CI?

LaurenzV commented:
Okay, I've figured the Carbon stuff out, only the match problem remaining. I've opened a PR, with that those tests pass locally for me: https://github.com/viperproject/silicon/pull/908

Once that's merged, we can rerun CI and hopefully merge this PR.

LaurenzV committed:
Merge branch 'master' into format-rnode

LaurenzV committed:
Re-run CI

LaurenzV committed:
Re-run CI

LaurenzV committed:
Merge branch 'master' into format-rnode

LaurenzV committed:
Merge branch 'master' into format-rnode

rayman2000 commented:
@JonasAlaif are you happy with this now? You still have "requested changes" but I think everything has been addressed.

rayman2000 committed:
Merge branch 'master' into format-rnode
